File: debug/src/main/java/com/sofar/debug/performance/monitor/FrameMonitor.java
package com.sofar.debug.performance.monitor;

import android.util.Log;
import android.view.Choreographer;

import com.sofar.debug.performance.PerformanceMonitorManager;

/**
 * 监控fps（Frames Per Second）
 */
public class FrameMonitor extends AbsMonitor {

  private static final String TAG = "FrameMonitor";

  private static final int FPS_SAMPLING_TIME = 1000;
  private static final int MAX_FRAME_RATE = 60;
  private int mLastFrameRate = MAX_FRAME_RATE;

  private FrameRateRunnable mRateRunnable = new FrameRateRunnable();

  private class FrameRateRunnable implements Runnable, Choreographer.FrameCallback {
    private int totalFramesPerSecond;

    @Override
    public void doFrame(long frameTimeNanos) {
      totalFramesPerSecond++;
      Choreographer.getInstance().postFrameCallback(this);
    }

    @Override
    public void run() {
      mLastFrameRate = totalFramesPerSecond;
      record();
      totalFramesPerSecond = 0;
      UIHandler.postDelayed(this, FPS_SAMPLING_TIME);
    }
  }


  @Override
  protected void onStart() {
    UIHandler.postDelayed(mRateRunnable, FPS_SAMPLING_TIME);
    Choreographer.getInstance().postFrameCallback(mRateRunnable);
  }

  @Override
  protected void onStop() {
    Choreographer.getInstance().removeFrameCallback(mRateRunnable);
    UIHandler.removeCallbacks(mRateRunnable);
  }

  private void record() {
    Log.d(TAG, "帧率=" + mLastFrameRate);
    PerformanceMonitorManager.get().frameCallback(mLastFrameRate);
  }

}

File: tool/profiler/src/main/java/com/sofar/profiler/frame/FrameMonitor.kt
package com.sofar.profiler.frame

import android.util.Log
import android.view.Choreographer
import com.sofar.profiler.AbsMonitor
import com.sofar.profiler.MonitorManager
import com.sofar.profiler.MonitorType

class FrameMonitor : AbsMonitor() {

  private var tag = "FrameMonitor"
  private val FPS_SAMPLING_TIME = 1000L
  private var frameRate = 60

  private val runnable = FrameRateRunnable()

  private inner class FrameRateRunnable : Runnable, Choreographer.FrameCallback {
    private var totalFramesPerSecond = 0
    override fun doFrame(frameTimeNanos: Long) {
      totalFramesPerSecond++
      Choreographer.getInstance().postFrameCallback(this)
    }

    override fun run() {
      frameRate = totalFramesPerSecond
      record()
      totalFramesPerSecond = 0
      UIHandler.postDelayed(this, FPS_SAMPLING_TIME)
    }
  }

  override fun onStart() {
    UIHandler.postDelayed(runnable, FPS_SAMPLING_TIME)
    Choreographer.getInstance().postFrameCallback(runnable)
  }

  override fun onStop() {
    Choreographer.getInstance().removeFrameCallback(runnable)
    UIHandler.removeCallbacks(runnable)
  }

  override fun type(): MonitorType {
    return MonitorType.FPS
  }

  fun record() {
    Log.d(tag, "fps=$frameRate")
    MonitorManager.frameCallback(frameRate)
  }

  fun getFPS(): Int {
    return frameRate
  }
}